### PR TITLE
Upgrade to relay v8

### PR DESCRIPTION
**Pull request by thicodes**
Adds support for relay 8

Related PR's:
- Remove DEPRECATE__create method (and fixes related to it)
  https://github.com/facebook/relay/commit/860c23cbb445e04fd14e5d05ed53a026d4f63357
- Flow [2/n] Generate an extra Local3DPayload opaque type for 3D fields
https://github.com/facebook/relay/commit/dafeea795a762811f087dcbaf026adf242bec171
- Fix generating Local3DPayload on some fields without @match
https://github.com/facebook/relay/commit/5cf60b351ecc249b3a3e2c085e115e9e26c10ba0
- Remove new connection model experiment
https://github.com/facebook/relay/commit/df17e69c2f07423f31d74d5c40e0b35273435688#diff-8612274e56204a0be25670395038c769

### CODE REVIEW

File: src/TypeScriptGenerator.ts [hunk 1]
```diff
@@ -776,17 +778,16 @@ function makeRawResponseProp(
         schema,
         [Array.from(nullthrows(nodeSelections).values())],
         state,
-        schema.isAbstractType(nodeType) ? null : nodeType.name
+        // TODO schema.isWrapper - when update @types/relay-compiler to v8 remove ts-ignore
+        // @ts-ignore
```
Comment (alloy) [inline src/TypeScriptGenerator.ts]: Can you make this update first?

File: src/TypeScriptGenerator.ts [hunk 2]
```diff
@@ -842,50 +843,64 @@ function selectionsToRawResponseBabel(
...
-              { ...selection, conditional: false },
-              state,
-              concreteType
-            );
-          }
-
-          return makeRawResponseProp(schema, selection, state, concreteType);
-        })
+        exactObjectTypeAnnotation(
+          // @ts-ignore
```
Comment (alloy) [inline src/TypeScriptGenerator.ts]: Why?

File: src/TypeScriptGenerator.ts [hunk 3]
```diff
@@ -842,50 +843,64 @@ function selectionsToRawResponseBabel(
...
-            { ...selection, conditional: false },
-            state,
-            nodeTypeName
-          );
-        }
-
-        return makeRawResponseProp(schema, selection, state, null);
-      })
+      exactObjectTypeAnnotation(
+        // @ts-ignore
```
Comment (alloy) [inline src/TypeScriptGenerator.ts]: Ditto

File: src/TypeScriptGenerator.ts [hunk 4]
```diff
@@ -842,50 +843,64 @@ function selectionsToRawResponseBabel(
...
+): void {
+  const moduleImport = selections.find(sel => sel.kind === "ModuleImport");
+  if (moduleImport) {
+    // Generate an extra opaque type for client 3D fields
+    state.runtimeImports.add("Local3DPayload");
+    types.push(
+      ts.createTypeReferenceNode(ts.createIdentifier("Local3DPayload"), [
+        stringLiteralTypeAnnotation(moduleImport.documentName as any),
+        exactObjectTypeAnnotation(
+          // @ts-ignore
```
Comment (alloy) [inline src/TypeScriptGenerator.ts]: Why?

File: src/TypeScriptGenerator.ts [hunk 5]
```diff
@@ -898,6 +913,7 @@ function createRawResponseTypeVisitor(
       Root(node) {
         return exportType(
           `${node.name}RawResponse`,
+          // @ts-ignore
```
Comment (alloy) [inline src/TypeScriptGenerator.ts]: Why?

File: src/TypeScriptGenerator.ts [hunk 6]
```diff
@@ -966,6 +987,7 @@ function createRawResponseTypeVisitor(
 }
 
 // Dedupe the generated type of module selections to reduce file size
+// @ts-ignore
```
Comment (alloy) [inline src/TypeScriptGenerator.ts]: Why?

File: src/TypeScriptGenerator.ts [hunk 7]
```diff
@@ -987,14 +1009,15 @@ function visitRawResponseModuleImport(
       null
     );
 
-    state.matchFields.set(key, ast);
+    state.matchFields.set(key, ast as any);
```
Comment (alloy) [inline src/TypeScriptGenerator.ts]: Why?

File: src/TypeScriptGenerator.ts [hunk 8]
```diff
@@ -1111,7 +1134,9 @@ function getEnumDefinitions(
   }
 
   return enumNames.map(name => {
-    const values = schema.getEnumValues(usedEnums[name]);
+    const values = [].concat(
+      schema.getEnumValues(usedEnums[name]) as any
+    ) as any;
```
Comment (alloy) [inline src/TypeScriptGenerator.ts]: Why?

File: src/TypeScriptGenerator.ts [hunk 9]
```diff
@@ -1121,6 +1146,7 @@ function getEnumDefinitions(
     return exportType(
       name,
       ts.createUnionTypeNode(
+        // @ts-ignore
```
Comment (alloy) [inline src/TypeScriptGenerator.ts]: Why?

File: test/TypeScriptGenerator-test.ts [hunk 10]
```diff
@@ -1,24 +1,19 @@
-import {
-  GraphQLCompilerContext,
-  IRTransforms,
-  Root,
-  transformASTSchema
-} from "relay-compiler";
+// @ts-ignore
+// TODO - Update CompilerContext type when release @types/relay-compiler 8
```
Comment (alloy) [inline test/TypeScriptGenerator-test.ts]: As before, can you make this update first?

**Comment by maraisr** [comment]
Without digging in too deep @thicodes ; does this also solve the removing of `thunks`?

**Comment by thicodes** [comment]
@maraisr Not yet

File: src/TypeScriptGenerator.ts (hunk 1 above)
Comment (thicodes) [reply src/TypeScriptGenerator.ts]: Yes. I started at https://github.com/DefinitelyTyped/DefinitelyTyped/pull/41740
When its ready I will ping you

Comment (alloy) [reply src/TypeScriptGenerator.ts]: Great, just merged it 👌

**Comment by alloy** [comment]
@thicodes Please let me know when this is good to review again 🙏

File: src/TypeScriptGenerator.ts (hunk 8 above)
Comment (thicodes) [reply src/TypeScriptGenerator.ts]: Concat is waiting a Array<never>. I fixed it by switching to spread in array

File: src/TypeScriptGenerator.ts (hunk 4 above)
Comment (thicodes) [reply src/TypeScriptGenerator.ts]: @alloy The problem is that exactObjectTypeAnnotation arg is `ts.PropertySignature[]`, but `selections` is `ReadonlyArray<Selection>`.
How could I fix this?

**Comment by maraisr** [comment]
Team @alloy @sibelius @renanmav now that v9 is released, we should probably start seriously looking at this PR.

**Comment by sibelius** [comment]
for me, we can merge with this `any`

and try to fix this on relay v9

**Comment by damassi** [comment]
Just ran into this issue trying to upgrade our project as well, but from `found-relay` which depends on Relay 9.

**Comment by alloy** [comment]
Superseded by https://github.com/relay-tools/relay-compiler-language-typescript/pull/173

**Pull request closed without merging**